### PR TITLE
Registry setup refactor

### PR DESCRIPTION
**Pull request by goral09**
## Overview
This is preliminary work for [RHOL-804](https://rchain.atlassian.net/browse/RHOL-804). It look like all the tasks are so connected together that it's really hard to create small PRs. 

This PR:
1. Refactors the `Registry` to remove circular dependency.
2. Refactors the `RholangAndScalaDispatcher.create` method.
3. Refactors the `PureRSpace` construction.
4. Introduces concept of `ChargingRSpace`.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
[RHOL-804](https://rchain.atlassian.net/browse/RHOL-804) 

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [ ] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [ ] You assigned one person to review this PR

### If you are not a member of the core development team, please confirm:
- [ ] You signed the commit. Merging requires a signature. Please see the [developer wiki](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain) for instructions.
- [ ] Your GitHub account is also an account with [Travis CI](https://travis-ci.org). Unit tests will not run on your PR unless you have an account with Travis. Merge requires passed unit tests.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala [hunk 1]
```diff
@@ -78,7 +78,7 @@ class ReduceSpec extends FlatSpec with Matchers with PersistentStoreTester {
     implicit val costAccounting =
       CostAccountingAlg.unsafe[Task](CostAccount(Integer.MAX_VALUE))
     val result = withTestSpace { space =>
-      val reducer      = RholangOnlyDispatcher.create[Task, Task.Par](space).reducer
+      val reducer      = RholangOnlyDispatcher.create[Task, Task.Par](space)._2
```
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Our tests seem to have a very nasty habit of duplicating test setup for all test cases...

Would that be too much of an ask if I asked you to make this test define the dispatcher once? I'm fine if it's using a `def` :)

I think it'd be best if we changed `withTestSpace { space =>` to `withTestSpaceAndReducer { space, reducer =>`

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala [hunk 2]
```diff
@@ -165,12 +157,6 @@ class Registry(private val space: Runtime.RhoPureSpace,
     Seq(Quote(Par(exprs = Seq(EVar(FreeVar(0))), connectiveUsed = true))),
     freeCount = 1)
 
-  private val insertCallbackRef: Long = Runtime.BodyRefs.REG_INSERT_CALLBACK
-
-  private val deleteRef: Long             = Runtime.BodyRefs.REG_DELETE
-  private val deleteRootCallbackRef: Long = Runtime.BodyRefs.REG_DELETE_ROOT_CALLBACK
-  private val deleteCallbackRef: Long     = Runtime.BodyRefs.REG_DELETE_CALLBACK
-
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala]: Those seem to be unrelated changes, please remove them from this commit - it's confusing as hell :)

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/RegistrySpec.scala [hunk 3]
```diff
@@ -26,6 +28,21 @@ trait RegistryTester extends PersistentStoreTester {
   implicit val errorLog = new ErrorLog()
   implicit val costAccounting =
     CostAccountingAlg.unsafe[Task](CostAccount(Integer.MAX_VALUE))
+
+  def dispatchTableCreator(registry: Registry): RhoDispatchMap =
```
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/RegistrySpec.scala]: This also is a change unrelated to this commit

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala [hunk 4]
```diff
@@ -109,20 +110,17 @@ class RholangAndScalaDispatcher[M[_]] private (
...
-  def create[M[_], F[_]](
-      tuplespace: RhoISpace,
-      dispatchTable: => Map[Long, Function1[Seq[ListChannelWithRandom], M[Unit]]],
-      urnMap: Map[String, Par])(implicit
-                                parallel: Parallel[M, F],
-                                s: Sync[M],
-                                ft: FunctorTell[M, Throwable])
-    : (Dispatch[M, ListChannelWithRandom, TaggedContinuation], Reduce[M]) = {
-    val pureSpace          = PureRSpace[M].of(tuplespace)
+  def create(tuplespace: RhoISpace,
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]:
```
 Fix `F[_]` to monix Task during creation of RholangAndScalaDispatcher.

This is necessary because Registry if fixed on Task and it seemed to me that
refactoring it to arbitrary effect would be much more time consuming. We never
really call RholangAndScalaDispatcher with anything different than Task anyway.
```

I fail to see how this is necessary (backed by 'seemed'). "It's only called with Task" is not a valid argument, for testability and maintainability reasons. If regarded a valid argument, it would justify throwing away all the tagless final effort.

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: Given that 'Registry is fixed on task', I also see that we've both done a very poor job when reviewing the Registry PR. I can only excuse us by the size of the PR - which itself was inexcusable :)

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: Instead of fixing Dispatcher to Task, please make the Registry abstract away from it. Change things for the better, not the worse. If you need that, I can do the Registry refactor for you.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Runtime.scala [hunk 5]
```diff
@@ -137,10 +137,10 @@ object Runtime {
     val errorLog                                  = new ErrorLog()
     implicit val ft: FunctorTell[Task, Throwable] = errorLog
 
-    def dispatchTableCreator(space: RhoISpace, dispatcher: RhoDispatch): RhoDispatchMap = {
+    def dispatchTableCreator(space: RhoISpace,
+                             dispatcher: RhoDispatch,
+                             registry: Registry): RhoDispatchMap = {
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Runtime.scala]: I like that you're unhardcoding the registry creation here.

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Runtime.scala]: Maybe `dispatchTableCreator` should be `createDispatchTable`?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+object ChargingRSpace {
+  def pureRSpace(implicit costAlg: CostAccountingAlg[Task], space: RhoPureSpace) =
+    new PureRSpace[Task,
+                   Channel,
+                   BindPattern,
+                   OutOfPhlogistonsError.type,
+                   ListChannelWithRandom,
+                   ListChannelWithRandom,
+                   TaggedContinuation] {
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala]: This object is unused and seems as if it needs some more love (use type aliases, remove commented out code, maybe elliminate some TODOs). It also lacks tests. Since it's not used yet, please remove this commit and propose the change with a PR that will use it.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala (hunk 1 above)
Comment (goral09) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: In few tests we use different reducer (`RholangAndScalaDispatcher`).

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala (hunk 2 above)
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala]: They are very much related and even necessary for everything to work. Because I moved where the `Registry` is being created (from the `dispatchTableCreator` to the `RholangAndScalaDispatcher.create` method) I couldn't use `testingDispatchTable` from the instance of `Registry`. I had to move it out from the instance to an object.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Runtime.scala (hunk 5 above)
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Runtime.scala]: I did it on purpose. See my previous comment.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/RegistrySpec.scala (hunk 3 above)
Comment (goral09) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/RegistrySpec.scala]: See my previous comments.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala (hunk 4 above)
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: I explained it in the commit description. Abstracting away `Registry`  from the `Task` is a time consuming change. We never call `RholangWithScalaDispatcher` in tests because it is not easily testable. That's way fixing it on `Task` is fine.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala (hunk 1 above)
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: I'm fine if you leave them out, and even more fine if you provide a separate method for them.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala (hunk 4 above)
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: The purpose of this PR is not to change the implementation of the `Registry`.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala (hunk 2 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala]: I still don't see how that makes it part of fixing the effect type to Task. This could be one or two separate commits. It'd help with changing the wrong part of the commit (fix over Task) by having it separate.

Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala]: You are now arguing that my git hygiene is bad. I know it can improve, I am working on it.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala (hunk 4 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: https://github.com/rchain/rchain/pull/1510

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala (hunk 2 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala]: I'm very pleased with your git hygiene in general and I appreciate that you're improving even now. The point of my comment was to point out how you can improve further :) And provide rationale why it might be worthy (b/c it'd make your and our work easier; in this particular case - b/c I'd like us to change the 'wrong' part of this commit)

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala [hunk 7]
```diff
@@ -28,14 +28,10 @@ import scala.collection.mutable.HashMap
...
+
 trait PersistentStoreTester {
-  def withTestSpace[R](
-      f: FreudianSpace[Channel,
-                       BindPattern,
-                       OutOfPhlogistonsError.type,
-                       ListChannelWithRandom,
-                       ListChannelWithRandom,
-                       TaggedContinuation] => R): R = {
+  def withTestSpace[R](errorLog: ErrorLog)(f: TestFixture => R): R = {
```
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala]: Would prefer just
`def withTestFixture[R](errorLog: ErrorLog)(f: (Rspace, Reducer) => R): R = {`
but that's mainly FYI - the duplication has been reduced :ok_hand:

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Runtime.scala (hunk 5 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Runtime.scala]: My point as a rename would be nice to have :confused:

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Runtime.scala]: Feel free to resolve this thread without the rename

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala (hunk 4 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/dispatch.scala]: Given #1510 is merged, there's no excuse for fixing over task. Please edit this commit so that it only touches the name registration part, and change the commit message. Notice also there's a likely merge conflict race between this and #1503 :) Which again reinforces the benefits of fine-grained commits.

**Comment by ArturGajowy** [comment]
Notice the outdated comments get folded too ;/

**Comment by goral09** [comment]
I think that is in the company's guidelines to merge with `dev` , not rebase.

**Comment by ArturGajowy** [comment]
I know. :)
